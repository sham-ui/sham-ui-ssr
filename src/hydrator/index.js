import Storage from './storage/service';
import ServerSideHooks from './hooks/service';
import ServerSideDOM from './dom/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Main function for setup hydrating components
 * @param {Object} DI App DI container
 */
export function setup( DI ) {
    const storage = new Storage();

    // Override default hooks service
    DI.bind(
        HOOKS_DI_KEY,

        // Pass default hooks service for proxy some methods (rehydrate & resolveID)
        new ServerSideHooks( DI.resolve( HOOKS_DI_KEY ), storage )
    );

    // Override default dom service
    DI.bind(
        DOM_DI_KEY,

        // Don't pass default dom service, because ServerSideDOM full override it
        new ServerSideDOM( storage )
    );

    return storage.root;
}


/**
 * Hydrate all components
 * @param {Object} DI App DI container
 * @return {Promise}
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import { setupHydrator, hydrate } from 'sham-ui-ssr/hydrator';
 *
 * const DI = createDI();
 * const root = setupHydrator( DI );
 * mainInitializer( DI, root );
 * start( DI );
 * hydrate( DI ).then( storage => {
 *     const { html, data } = storage.hydrate();
 * } )
 */
export function hydrate( DI ) {
    return DI.resolve( HOOKS_DI_KEY ).hydrating();
}
