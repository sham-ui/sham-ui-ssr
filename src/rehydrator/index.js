import RestoreDom from './dom/service';
import RehydrateHooks from './hooks/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Save hooks & dom services for restore after rehydrating
 * @param {Object} DI App DI container
 * @return {Function}
 */
function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

/**
 * Main function for rehydrating
 * @param {Object} DI App DI container
 * @param {Object} storage
 * @return {Function} Restore function for disable rehydrating and switch to default behavior
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import setupUnsafe from 'sham-ui-unsafe';
 * import { setupRehydrator } from 'sham-ui-ssr/rehydrator';
 *
 * const DI = createDI();
 * const data = JSON.parse( document.getElementById( 'data' ).innerText );
 * const disableRehydrating = setupRehydrator( DI, data );
 * mainInitializer( DI, root );
 * start( DI );
 * disableRehydrating();
 * setupUnsafe( DI );
 */
export function setup( DI, storage ) {
    const restore = makeRestore( DI );
    DI.bind( HOOKS_DI_KEY, new RehydrateHooks( storage ) );
    DI.bind( DOM_DI_KEY, new RestoreDom( storage ) );
    return restore;
}
