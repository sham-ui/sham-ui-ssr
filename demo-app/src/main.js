import './styles/main.scss';
import setupUnsafe from 'sham-ui-unsafe';
import { createDI, start } from 'sham-ui';
import { setup as setupHydrator, hydrate } from '../../src/hydrator/index';
import { setup as setupRehydrator } from '../../src/rehydrator/index';
import mainInitializer from './initializers/main';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

const DI = createDI();
const restoreOriginal = makeRestore( DI );

const root = setupHydrator( DI );
mainInitializer( DI, root );
start( DI );
hydrate( DI ).then(
    ( storage ) => {
        const getById = document.getElementById.bind( document );
        // eslint-disable-next-line no-console
        console.dir( storage );
        try {
            const { html, data } = storage.hydrate();
            getById( 'ssr-rendered' ).innerHTML = html;
            getById( 'raw-html' ).innerText = html;
            getById( 'data' ).innerText = JSON.stringify( JSON.parse( data ), null, 4 );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
        try {
            DI.resolve( 'sham-ui:store' ).byId.clear();
            const root = getById( 'client' );
            root.innerHTML = getById( 'ssr-rendered' ).innerHTML;
            restoreOriginal();
            const data = JSON.parse( document.getElementById( 'data' ).innerText );
            const disableRehydrating = setupRehydrator( DI, data );
            mainInitializer( DI, root );
            start( DI );
            disableRehydrating();
            setupUnsafe( DI );
            // eslint-disable-next-line no-console
            console.dir( DI.resolve( 'sham-ui:store' ).byId.keys() );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
    }
);

