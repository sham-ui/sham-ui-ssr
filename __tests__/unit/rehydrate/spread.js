import { compile } from 'sham-ui-test-helpers';
import { ssrAndRehydrate } from '../../../src/testing';

beforeEach( () => {
    window.SpreadCustom = compile`
        <i>{{ foo }}</i>
        <i>{{ boo }}</i>
        <i>{{ bar }}</i>
    `;
} );

afterEach( () => {
    delete window.SpreadCustom;
} );

it( 'should render arrays', async() => {
    expect.assertions( 4 );
    const meta = await ssrAndRehydrate(
        compile`
            <ul>
                {% for key, value of list %}
                    <li>{{ key }}:{{ value }}</li>
                {% endfor %}
            </ul>
        `,
        {
            list: [ 1, 2, 3 ]
        }
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<ul><li>0<!--0-->:<!--1-->1</li><li>1<!--0-->:<!--1-->2</li><li>2<!--0-->:<!--1-->3</li></ul>'
    );

    meta.component.update( { list: [ 1, 3 ] } );
    expect( meta.component.container.innerHTML ).toBe(
        '<ul><li>0<!--0-->:<!--1-->1</li><li>1<!--0-->:<!--1-->3</li></ul>'
    );

    meta.component.update( { list: [ 'a', 'b', 'c', 'd' ] } );
    expect( meta.component.container.innerHTML ).toBe(
        '<ul><li>0<!--0-->:<!--1-->a</li><li>1<!--0-->:<!--1-->b</li><li>2:c</li><li>3:d</li></ul>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work for html elements', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`
            <div {{...attr}}></div>
        `,
        {
            attr: {
                id: 'id',
                'data-attr': 'data',
                'class': 'foo'
            }
        }
    );
    expect( meta.html ).toBe( '<div id="id" data-attr="data" class="foo"></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should override default attributes', async() => {
    expect.assertions( 3 );
    const meta = await ssrAndRehydrate(
        compile`<div id="foo" {{...attr}}></div>`
    );
    expect( meta.html ).toBe( '<div id="foo"></div>' );

    meta.component.update( {
        attr: {
            id: 'boo'
        }
    } );
    expect( meta.component.container.innerHTML ).toBe( '<div id="boo"></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should override variables attributes', async() => {
    expect.assertions( 4 );
    const meta = await ssrAndRehydrate(
        compile`
            <div id="{{ id }}" {{...attr}}></div>
        `,
        {
            id: 'foo'
        }
    );
    expect( meta.html ).toBe( '<div id="foo"></div>' );

    meta.component.update( {
        attr: {
            id: 'boo'
        }
    } );
    expect( meta.component.container.innerHTML ).toBe( '<div id="boo"></div>' );

    meta.component.update( { id: 'bar', attr: {} } );
    expect( meta.component.container.innerHTML ).toBe( '<div id="bar"></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work for custom tags', async() => {
    expect.assertions( 3 );
    const meta = await ssrAndRehydrate(
        compile`
            <div>
                <SpreadCustom {{...attr}}/>
            </div>
        `,
        {
            attr: {
                foo: 'foo',
                boo: 'boo',
                bar: 'bar'
            }
        }
    );
    expect( meta.html ).toBe( '<div><i>foo</i><i>boo</i><i>bar</i></div>' );

    meta.component.update( {
        attr: {
            boo: 'Boo-Ya'
        }
    } );
    expect( meta.component.container.innerHTML ).toBe(
        // eslint-disable-next-line max-len
        '<div><i>foo</i><i>Boo-Ya</i><i>bar</i></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work for custom tags with constant attributes values', async() => {
    expect.assertions( 4 );
    const meta = await ssrAndRehydrate(
        compile`
            <div>
                <SpreadCustom {{...attr}} foo="foo"/>
            </div>
        `
    );
    expect( meta.html ).toBe( '<div><i>foo</i><i> </i><i> </i></div>' );

    meta.component.update( {
        attr: {
            boo: 'boo',
            bar: 'bar'
        }
    } );
    expect( meta.component.container.innerHTML ).toBe(
        '<div><i>foo</i><i>boo</i><i>bar</i></div>'
    );

    meta.component.update( {
        attr: {
            foo: 'over foo'
        }
    } );
    expect( meta.component.container.innerHTML ).toBe(
        '<div><i>over foo</i><i>boo</i><i>bar</i></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work for custom tags with attributes with values', async() => {
    expect.assertions( 5 );
    const meta = await ssrAndRehydrate(
        compile`
            <div>
                <SpreadCustom {{...attr}} foo="{{ foo }}"/>
            </div>
        `
    );
    expect( meta.html ).toBe( '<div></div>' );

    meta.component.update( {
        attr: {
            boo: 'boo',
            bar: 'bar'
        }
    } );
    expect( meta.component.container.innerHTML ).toBe( '<div><i></i><i>boo</i><i>bar</i></div>' );

    meta.component.update( {
        foo: 'foo'
    } );
    expect( meta.component.container.innerHTML ).toBe(
        '<div><i>foo</i><i>boo</i><i>bar</i></div>'
    );

    meta.component.update( {
        attr: {
            foo: 'over foo'
        },
        foo: 'for'
    } );
    expect( meta.component.container.innerHTML ).toBe(
        '<div><i>for</i><i>boo</i><i>bar</i></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work for custom tags proxy __data__', async() => {
    expect.assertions( 3 );
    const meta = await ssrAndRehydrate(
        compile`
            <div>
                <SpreadCustom {{...this.__data__}}/>
            </div>
        `,
        {
            foo: 'foo',
            boo: 'boo',
            bar: 'bar'
        }
    );
    expect( meta.html ).toBe( '<div><i>foo</i><i>boo</i><i>bar</i></div>' );

    meta.component.update( {
        boo: 'Boo-Ya'
    } );
    expect( meta.component.container.innerHTML ).toBe(
        '<div><i>foo</i><i>Boo-Ya</i><i>bar</i></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
