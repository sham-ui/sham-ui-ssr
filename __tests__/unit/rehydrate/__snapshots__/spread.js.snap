// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should override default attributes 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "attr": Object {
      "id": "boo",
    },
  },
  "html": "<div id=\\"boo\\"></div>",
}
`;

exports[`should override variables attributes 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "attr": Object {},
    "id": "bar",
  },
  "html": "<div id=\\"bar\\"></div>",
}
`;

exports[`should render arrays 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "list": Array [
      "a",
      "b",
      "c",
      "d",
    ],
  },
  "html": "
<ul>
  <li>0
    <!--0-->:
    <!--1-->a
  </li>
  <li>1
    <!--0-->:
    <!--1-->b
  </li>
  <li>2:c</li>
  <li>3:d</li>
</ul>
",
}
`;

exports[`should work for custom tags 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "attr": Object {
      "boo": "Boo-Ya",
    },
  },
  "html": "
<div>
  <i>foo</i>
  <i>Boo-Ya</i>
  <i>bar</i>
</div>
",
}
`;

exports[`should work for custom tags proxy __data__ 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "bar": "bar",
    "boo": "Boo-Ya",
    "foo": "foo",
  },
  "html": "
<div>
  <i>foo</i>
  <i>Boo-Ya</i>
  <i>bar</i>
</div>
",
}
`;

exports[`should work for custom tags with attributes with values 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "attr": Object {
      "foo": "over foo",
    },
    "foo": "for",
  },
  "html": "
<div>
  <i>for</i>
  <i>boo</i>
  <i>bar</i>
</div>
",
}
`;

exports[`should work for custom tags with constant attributes values 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "attr": Object {
      "foo": "over foo",
    },
  },
  "html": "
<div>
  <i>over foo</i>
  <i>boo</i>
  <i>bar</i>
</div>
",
}
`;

exports[`should work for html elements 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "attr": Object {
      "class": "foo",
      "data-attr": "data",
      "id": "id",
    },
  },
  "html": "<div id=\\"id\\" data-attr=\\"data\\" class=\\"foo\\"></div>",
}
`;
